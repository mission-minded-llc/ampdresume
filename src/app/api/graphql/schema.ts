import { makeExecutableSchema } from "@graphql-tools/schema";
import { PrismaClient } from "@prisma/client";
import gql from "graphql-tag";

const prisma = new PrismaClient();

// Define type definitions
const typeDefs = gql`
  type Query {
    # General queries.
    skills: [Skill!]!

    # User-specific queries.
    user(slug: String!): User!
    skillsForUser(userId: ID!): [SkillForUser!]!
    companies(userId: ID!, sort: [SortInput!]): [Company!]!
    positions(companyIds: [ID!], sort: [SortInput!]): [Position!]!
    education(userId: ID!, sort: [SortInput!]): [Education!]!
  }

  type Mutation {
    # User-specific mutations.
    addSkillForUser(userId: ID!, skillId: ID!, yearStarted: Int, totalYears: Int): SkillForUser!
  }

  type User {
    id: ID!
    name: String!
    email: String!
    displayEmail: String
    location: String
    title: String
    siteTitle: String
    siteImage: String
  }

  type SkillsForUser {
    skillsForUser: [SkillForUser!]
  }

  type SkillForUser {
    id: ID!
    skill: Skill!
    icon: String
    description: String
    yearStarted: Int
    totalYears: Int
  }

  type Skill {
    id: ID!
    name: String!
    icon: String
  }

  type Company {
    id: ID!
    name: String!
    location: String
    startDate: String
    endDate: String
  }

  type Position {
    id: ID!
    title: String!
    startDate: String
    endDate: String
    company: Company!
    projects: [Project!]!
  }

  type Project {
    id: ID!
    name: String!
    description: String
    skillsForProject: [SkillForProject!]!
  }

  type SkillForProject {
    id: ID!
    skillForUser: SkillForUser!
    description: String
  }

  type Education {
    id: ID!
    school: String!
    degree: String
    dateAwarded: String
  }

  input SortInput {
    field: String!
    direction: SortDirection!
  }

  enum SortDirection {
    ASC
    DESC
  }
`;

const resolvers = {
  Query: {
    // General queries.
    skills: async () => {
      return await prisma.skill.findMany();
    },

    // User-specific queries.
    user: async (_: string, { slug }: { slug: string }) => {
      return await prisma.user.findUnique({
        where: { slug },
      });
    },
    skillsForUser: async (_: string, { userId }: { userId: string }) => {
      return await prisma.skillForUser.findMany({
        where: { userId },
        include: { skill: true }, // Include skill details
      });
    },
    companies: async (
      _: string,
      {
        userId,
        sort,
      }: { userId: string; sort: Array<{ field: string; direction: "ASC" | "DESC" }> },
    ) => {
      // Map the sort array into Prisma-compatible orderBy
      const orderBy =
        sort?.map(({ field, direction }) => ({
          [field]: direction.toLowerCase(), // Prisma expects lowercase for ASC/DESC
        })) || [];

      return await prisma.company.findMany({
        where: { userId },
        orderBy, // Apply sorting
      });
    },
    positions: async (
      _: string,
      {
        companyIds,
        sort,
      }: { companyIds: string[]; sort: Array<{ field: string; direction: "ASC" | "DESC" }> },
    ) => {
      // Map the sort array into Prisma-compatible orderBy
      const orderBy =
        sort?.map(({ field, direction }) => ({
          [field]: direction.toLowerCase(), // Prisma expects lowercase for ASC/DESC
        })) || [];

      return await prisma.position.findMany({
        where: { companyId: { in: companyIds } },
        orderBy, // Apply sorting
        include: {
          company: true, // Include company details
          projects: { include: { skillsForProject: { include: { skillForUser: true } } } },
        }, // Include project and skill details
      });
    },
    education: async (
      _: string,
      {
        userId,
        sort,
      }: { userId: string; sort: Array<{ field: string; direction: "ASC" | "DESC" }> },
    ) => {
      // Map the sort array into Prisma-compatible orderBy
      const orderBy =
        sort?.map(({ field, direction }) => ({
          [field]: direction.toLowerCase(), // Prisma expects lowercase for ASC/DESC
        })) || [];

      return await prisma.education.findMany({
        where: { userId },
        orderBy, // Apply sorting
      });
    },
  },

  // Mutation resolvers.
  Mutation: {
    // User-specific mutations.
    addSkillForUser: async (
      _: string,
      {
        userId,
        skillId,
        yearStarted,
        totalYears,
      }: { userId: string; skillId: string; yearStarted: number; totalYears: number },
    ) => {
      return await prisma.skillForUser.create({
        data: {
          userId,
          skillId,
          yearStarted,
          totalYears,
        },
        include: { skill: true }, // Include skill details
      });
    },
  },
};

export const schema = makeExecutableSchema({ typeDefs, resolvers });
