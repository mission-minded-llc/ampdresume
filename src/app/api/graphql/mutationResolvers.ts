import { PrismaClient } from "@prisma/client";

const prisma = new PrismaClient();

export const mutationResolvers = {
  // User-specific mutations.
  addSkillForUser: async (
    _: string,
    {
      userId,
      skillId,
      yearStarted,
      totalYears,
    }: { userId: string; skillId: string; yearStarted: number; totalYears: number },
  ) => {
    // Check if this skill already exists for this user.
    const existingSkill = await prisma.skillForUser.findFirst({
      where: { userId, skillId },
    });

    if (existingSkill) {
      return {
        id: existingSkill.id,
      };
    }

    return await prisma.skillForUser.create({
      data: {
        userId,
        skillId,
        yearStarted,
        totalYears,
      },
      include: { skill: true }, // Include skill details
    });
  },

  updateSkillForUser: async (
    _: string,
    {
      id,
      userId,
      yearStarted,
      totalYears,
      description,
    }: {
      id: string;
      userId: string;
      yearStarted: number;
      totalYears: number;
      description: string;
    },
  ) => {
    /// TODO: verify that userId matches the current session user.

    return await prisma.skillForUser.update({
      where: { id },
      data: {
        userId,
        description,
        yearStarted,
        totalYears,
      },
      include: { skill: true }, // Include skill details
    });
  },
};
