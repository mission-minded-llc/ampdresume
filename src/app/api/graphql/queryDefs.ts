import { gql } from "@apollo/client";

export const queryDefs = gql`
  type Query {
    # General queries.
    skills: [Skill!]!

    # User-specific queries.
    user(slug: String!): User!
    skillsForUser(userId: ID!): [SkillForUser!]!
    companies(userId: ID!, sort: [SortInput!]): [Company!]!
    positions(companyIds: [ID!], sort: [SortInput!]): [Position!]!
    education(userId: ID!, sort: [SortInput!]): [Education!]!
  }
`;
