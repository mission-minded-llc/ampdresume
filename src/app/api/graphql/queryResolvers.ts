import { prisma } from "@/lib/prisma";

export const queryResolvers = {
  // General queries.
  skills: async () => {
    return await prisma.skill.findMany();
  },

  // User-specific queries.
  user: async (_: string, { slug }: { slug: string }) => {
    return await prisma.user.findUnique({
      where: { slug },
    });
  },

  skillsForUser: async (_: string, { userId }: { userId: string }) => {
    return await prisma.skillForUser.findMany({
      where: { userId },
      include: { skill: true }, // Include skill details
    });
  },

  companies: async (
    _: string,
    { userId, sort }: { userId: string; sort: Array<{ field: string; direction: "ASC" | "DESC" }> },
  ) => {
    // Map the sort array into Prisma-compatible orderBy
    const orderBy =
      sort?.map(({ field, direction }) => ({
        [field]: direction.toLowerCase(), // Prisma expects lowercase for ASC/DESC
      })) || [];

    return await prisma.company.findMany({
      where: { userId },
      orderBy, // Apply sorting
    });
  },

  positions: async (
    _: string,
    {
      companyIds,
      sort,
    }: { companyIds: string[]; sort: Array<{ field: string; direction: "ASC" | "DESC" }> },
  ) => {
    // Map the sort array into Prisma-compatible orderBy
    const orderBy =
      sort?.map(({ field, direction }) => ({
        [field]: direction.toLowerCase(), // Prisma expects lowercase for ASC/DESC
      })) || [];

    return await prisma.position.findMany({
      where: { companyId: { in: companyIds } },
      orderBy, // Apply sorting
      include: {
        company: true, // Include company details
        projects: { include: { skillsForProject: { include: { skillForUser: true } } } },
      }, // Include project and skill details
    });
  },

  education: async (
    _: string,
    { userId, sort }: { userId: string; sort: Array<{ field: string; direction: "ASC" | "DESC" }> },
  ) => {
    // Map the sort array into Prisma-compatible orderBy
    const orderBy =
      sort?.map(({ field, direction }) => ({
        [field]: direction.toLowerCase(), // Prisma expects lowercase for ASC/DESC
      })) || [];

    return await prisma.education.findMany({
      where: { userId },
      orderBy, // Apply sorting
    });
  },
};
