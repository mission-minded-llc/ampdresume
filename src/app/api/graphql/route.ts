import { NextRequest } from "next/server";
import { createYoga } from "graphql-yoga";
import { schema } from "@/graphql/server/schema";
import { getServerSession } from "next-auth";
import { authOptions } from "@/lib/auth";

const yoga = createYoga<NextRequest>({
  schema,
  graphqlEndpoint: "/api/graphql",
  fetchAPI: { Response: Response }, // Map Yoga's Response to the Next.js Response
  context: async () => {
    const session = await getServerSession(authOptions);
    return {
      session,
      isAuthenticated: Boolean(session?.user?.id),
      userId: session?.user?.id,
    };
  },
  // Add some basic security measures
  cors: {
    origin:
      process.env.NODE_ENV === "production"
        ? [process.env.NEXTAUTH_URL || "https://ampdresume.com"]
        : ["http://localhost:3000"],
    credentials: true,
  },
});

export const GET = async (request: NextRequest) => {
  return yoga(request);
};

export const POST = async (request: NextRequest) => {
  return yoga(request);
};
