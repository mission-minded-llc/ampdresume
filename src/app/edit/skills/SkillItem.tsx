import { SkillForUser } from "@ampdresume/theme";
import { Icon } from "@iconify/react";
import { Box, Dialog, DialogContent } from "@mui/material";
import Button from "@mui/material/Button";
import { usePathname } from "next/navigation";
import { useSession } from "next-auth/react";
import React, { useState } from "react";

import { SkillItemEdit } from "@/app/edit/skills/SkillItemEdit";
import { CustomDialogTitle } from "@/components/CustomDialogTitle";

export const SkillItem = ({ skill }: { skill: SkillForUser }) => {
  const { data: session, status } = useSession();
  const pathname = usePathname();

  const [isOpen, setIsOpen] = useState(false);
  const [icon, setIcon] = useState<string | null | undefined>(
    skill?.icon || skill?.skill?.icon || null,
  );

  const userCanEdit =
    skill?.userId &&
    pathname.startsWith("/edit/skills") &&
    status === "authenticated" &&
    session?.user?.id === skill.userId;

  const buttonDisabled = !(skill?.description || userCanEdit);

  const SkillIcon = () => (icon ? <Icon icon={icon} /> : null);

  return (
    <React.Fragment>
      <Button
        disabled={buttonDisabled}
        variant="outlined"
        color="primary"
        onClick={() => setIsOpen(true)}
        sx={(theme) => ({
          padding: "2px 10px",
          color: theme.palette.primary.main,
          backgroundColor: theme.palette.primary.light,
          boxShadow: `2px 2px 3px 0px ${theme.palette.primary.dark}`,
          "&.Mui-disabled": {
            color: theme.palette.primary.main,
            borderColor: "transparent",
            boxShadow: "none",
          },
          textTransform: "none",
          gap: "8px",
          borderColor: userCanEdit && skill?.description ? "lawngreen" : theme.palette.primary.dark,
        })}
      >
        <SkillIcon />
        {skill.skill.name}
      </Button>

      <Dialog open={isOpen} onClose={() => setIsOpen(false)} fullWidth maxWidth="md">
        <CustomDialogTitle closeHandler={() => setIsOpen(false)}>
          <Box sx={{ display: "flex", alignItems: "center", gap: "1em" }}>
            <SkillIcon />
            {skill.skill.name}
          </Box>
        </CustomDialogTitle>
        <DialogContent sx={{ padding: { xs: 2, sm: 3 } }}>
          <SkillItemEdit
            skill={skill}
            handleClose={() => setIsOpen(false)}
            setIconCallback={setIcon}
          />
        </DialogContent>
      </Dialog>
    </React.Fragment>
  );
};
