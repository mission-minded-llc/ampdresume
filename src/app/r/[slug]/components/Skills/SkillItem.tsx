import { Dialog, DialogContent, DialogTitle, IconButton } from "@mui/material";
import React, { useState } from "react";

import Button from "@mui/material/Button";
import { Icon } from "@iconify/react";
import styles from "./SkillItem.module.scss";
import { SkillForUserWithSkill } from "@/graphql/getSkillsForUser";

export const SkillItem = ({ skill }: { skill: SkillForUserWithSkill }) => {
  const [open, setOpen] = useState(false);

  return (
    <React.Fragment>
      <Button
        disabled={!skill?.description}
        variant="outlined"
        color="primary"
        className={styles.item}
        onClick={() => setOpen(true)}
        sx={(theme) => ({
          color: theme.palette.primary.main,
          "&.Mui-disabled": {
            color: theme.palette.primary.main,
            borderColor: "transparent",
          },
        })}
      >
        {skill?.icon ? (
          <Icon icon={skill.icon} className={styles.icon} />
        ) : skill?.skill?.icon ? (
          <Icon icon={skill.skill.icon} className={styles.icon} />
        ) : null}
        {skill.skill.name}
      </Button>
      <Dialog open={open} onClose={() => setOpen(false)}>
        <DialogTitle
          sx={{ display: "flex", alignItems: "center", gap: "1em", padding: "16px 48px" }}
        >
          {skill?.icon ? <Icon icon={skill.icon} className={styles.icon} /> : null}
          {skill.skill.name}
        </DialogTitle>
        <IconButton
          aria-label="close"
          onClick={() => setOpen(false)}
          sx={(theme) => ({
            position: "absolute",
            right: 8,
            top: 8,
            color: theme.palette.grey[500],
          })}
        >
          X
        </IconButton>
        <DialogContent>{skill.description}</DialogContent>
      </Dialog>
    </React.Fragment>
  );
};
