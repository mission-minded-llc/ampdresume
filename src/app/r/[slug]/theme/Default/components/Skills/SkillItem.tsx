import { Box, Dialog, DialogContent } from "@mui/material";
import React, { useContext, useState } from "react";

import Button from "@mui/material/Button";
import { CustomDialogTitle } from "@/components/DialogTitle";
import { Icon } from "@iconify/react";
import { SkillForProjectWithSkill } from "@/graphql/getSkillsForProject";
import { SkillForUserWithSkill } from "@/graphql/getSkillsForUser";
import { SkillItemEdit } from "@/app/edit/skills/SkillItemEdit";
import { SkillItemView } from "./SkillItemView";
import { SkillsContext } from "./Skills";
import { usePathname } from "next/navigation";
import { useSession } from "next-auth/react";

export const SkillItem = ({
  skill,
}: {
  skill: SkillForUserWithSkill | SkillForProjectWithSkill;
}) => {
  const { skillType } = useContext(SkillsContext);
  const { data: session, status } = useSession();
  const pathname = usePathname();

  const [isOpen, setIsOpen] = useState(false);

  const projectSkill = skill as SkillForProjectWithSkill;
  const skillData =
    skillType === "project"
      ? { ...projectSkill.skillForUser, description: projectSkill.description }
      : (skill as SkillForUserWithSkill);

  const userCanEdit =
    skillData?.userId &&
    pathname.startsWith("/edit/skills") &&
    status === "authenticated" &&
    session?.user?.id === skillData.userId;

  const buttonDisabled = !(skill?.description || userCanEdit);

  const SkillIcon = () =>
    skillData?.icon ? (
      <Icon icon={skillData.icon} />
    ) : skillData?.skill?.icon ? (
      <Icon icon={skillData.skill.icon} />
    ) : null;

  return (
    <React.Fragment>
      <Button
        disabled={buttonDisabled}
        variant="outlined"
        color="primary"
        onClick={() => setIsOpen(true)}
        sx={(theme) => ({
          padding: "2px 10px",
          color: theme.palette.primary.main,
          backgroundColor: theme.palette.primary.light,
          boxShadow: `2px 2px 3px 0px ${theme.palette.primary.dark}`,
          "&.Mui-disabled": {
            color: theme.palette.primary.main,
            borderColor: "transparent",
            boxShadow: "none",
          },
          textTransform: "none",
          gap: "8px",
          borderColor: userCanEdit && skill?.description ? "lawngreen" : theme.palette.primary.dark,
        })}
      >
        {skillData?.icon ? (
          <Icon icon={skillData.icon} />
        ) : skillData?.skill?.icon ? (
          <Icon icon={skillData.skill.icon} />
        ) : null}
        {skillData.skill.name}
      </Button>

      <Dialog open={isOpen} onClose={() => setIsOpen(false)} fullWidth maxWidth="md">
        <CustomDialogTitle closeHandler={() => setIsOpen(false)}>
          <Box sx={{ display: "flex", alignItems: "center", gap: "1em" }}>
            <SkillIcon />
            {skillData.skill.name}
          </Box>
        </CustomDialogTitle>
        <DialogContent>
          {userCanEdit ? (
            <SkillItemEdit skill={skillData} handleClose={() => setIsOpen(false)} />
          ) : (
            <SkillItemView skill={skillData} />
          )}
        </DialogContent>
      </Dialog>
    </React.Fragment>
  );
};
