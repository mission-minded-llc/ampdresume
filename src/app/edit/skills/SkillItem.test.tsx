import "@testing-library/jest-dom";

import { fireEvent, render, waitFor } from "@testing-library/react";

import React from "react";
import { SkillForUserWithSkill } from "@/graphql/getSkillsForUser";
import { SkillItem } from "./SkillItem";
import { usePathname } from "next/navigation";
import { useSession } from "next-auth/react";

jest.mock("next-auth/react", () => ({
  useSession: jest.fn(),
}));

jest.mock("next/navigation", () => ({
  usePathname: jest.fn(),
}));

jest.mock("@/app/edit/skills/SkillItemEdit", () => ({
  SkillItemEdit: ({
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    skill,
    handleClose,
  }: {
    skill: SkillForUserWithSkill;
    handleClose: () => void;
  }) => (
    <div>
      SkillItemEdit
      <button onClick={handleClose}>Close</button>
    </div>
  ),
}));

describe("SkillItem", () => {
  const mockSkill: SkillForUserWithSkill = {
    id: "1",
    userId: "user-id",
    skillId: "skill-id",
    yearStarted: 2020,
    totalYears: 2,
    skill: { id: "skill-id", name: "JavaScript", icon: "icon-js" },
    description: "Skill description",
    icon: "icon-js",
  };

  const mockSession = {
    user: { id: "user-id" },
  };

  beforeEach(() => {
    jest.clearAllMocks();
    (useSession as jest.Mock).mockReturnValue({ data: mockSession, status: "authenticated" });
    (usePathname as jest.Mock).mockReturnValue("/edit/skills");
  });

  it("renders correctly", () => {
    const { getByText } = render(<SkillItem skill={mockSkill} />);
    expect(getByText("JavaScript")).toBeInTheDocument();
  });

  it("opens the dialog when the button is clicked", async () => {
    const { getByText, getByRole } = render(<SkillItem skill={mockSkill} />);
    const button = getByText("JavaScript");

    fireEvent.click(button);

    await waitFor(() => {
      expect(getByRole("dialog")).toBeInTheDocument();
      expect(getByText("SkillItemEdit")).toBeInTheDocument();
    });
  });

  it("closes the dialog when the close button is clicked", async () => {
    const { container, getByText, getByRole, queryByRole } = render(
      <SkillItem skill={mockSkill} />,
    );
    const button = getByText("JavaScript");
    expect(container).toMatchSnapshot();

    fireEvent.click(button);

    await waitFor(() => {
      expect(getByRole("dialog")).toBeInTheDocument();
      expect(container).toMatchSnapshot();
    });

    fireEvent.click(getByText("Close"));

    await waitFor(() => {
      expect(queryByRole("dialog")).not.toBeInTheDocument();
      expect(container).toMatchSnapshot();
    });
  });
});
