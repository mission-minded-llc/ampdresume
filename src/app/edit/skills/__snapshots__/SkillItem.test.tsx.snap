// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`SkillItem closes the dialog when the close button is clicked 1`] = `
<div>
  <button
    class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary css-bxhczq-MuiButtonBase-root-MuiButton-root"
    tabindex="0"
    type="button"
  >
    <span
      data-icon="icon-js"
      data-testid="icon"
    />
    JavaScript
  </button>
</div>
`;

exports[`SkillItem closes the dialog when the close button is clicked 2`] = `
<div
  aria-hidden="true"
>
  <button
    class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary css-bxhczq-MuiButtonBase-root-MuiButton-root"
    tabindex="0"
    type="button"
  >
    <span
      data-icon="icon-js"
      data-testid="icon"
    />
    JavaScript
  </button>
</div>
`;

exports[`SkillItem closes the dialog when the close button is clicked 3`] = `
<div>
  <button
    class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary css-bxhczq-MuiButtonBase-root-MuiButton-root"
    tabindex="0"
    type="button"
  >
    <span
      data-icon="icon-js"
      data-testid="icon"
    />
    JavaScript
  </button>
</div>
`;
